### ISSUE TITLE
buildah run/bud --isolation=chroot $c <cmd> fails with single uidmap for unprivileged user

### ISSUE THREAD

**Issue by mrinaldhillon** [containers/buildah#2459]
buildah run/bud **fails with chroot isolation** if user is not configured in subuid/subgid map files. It also ***fails with oci isolation and crun runtime**. However **succeeds with rootless isolation with crun runtime**. Is this expected ? 

Fail: isolation=chroot
Fail: isolation=oci runtime=crun
Success: isolation=rootless runtime=crun

**Steps to reproduce the issue:**
1. remove user entries from /etc/subuid /etc/subgid
2. c=$(buildah from debian:10-slim)
3. buildah run --isolation=chroot $c true 

**Describe the results you received:**
```
buildah run --isolation=chroot $c true
error clearing supplemental groups list: operation not permittedsubprocess exited with status 1
                                                                                               exit status 1
```

```
buildah run --isolation=oci --runtime=crun $c true
2020-07-12T03:34:45.000749013Z: mount `devpts` to '/dev/pts': Invalid argument
error running container: error creating container for [/bin/true]: : exit status 1
error while running runtime: exit status 1
ERRO exit status 1
```

```
buildah run --isolation=rootless --runtime=crun $c true
echo $?
0
```

**Describe the results you expected:**
 buildah run --isolation=chroot $c true
echo $?
0 

**Output of `rpm -q buildah` or `apt list buildah`:**

```
buildah-1.15.0-1.fc32.x86_64
```

**Output of `buildah version`:**

```
Version:         1.15.0
Go Version:      go1.14.3
Image Spec:      1.0.1-dev
Runtime Spec:    1.0.2-dev
CNI Spec:        0.4.0
libcni Version:
image Version:   5.5.1
Git Commit:
Built:           Wed Dec 31 19:00:00 1969
OS/Arch:         linux/amd64
```

**Output of `cat /etc/*release`:**

```
Fedora release 32 (Thirty Two)
NAME=Fedora
VERSION="32 (Workstation Edition)"
ID=fedora
VERSION_ID=32
VERSION_CODENAME=""
PLATFORM_ID="platform:f32"
PRETTY_NAME="Fedora 32 (Workstation Edition)"
ANSI_COLOR="0;34"
LOGO=fedora-logo-icon
CPE_NAME="cpe:/o:fedoraproject:fedora:32"
HOME_URL="https://fedoraproject.org/"
DOCUMENTATION_URL="https://docs.fedoraproject.org/en-US/fedora/f32/system-administrators-guide/"
SUPPORT_URL="https://fedoraproject.org/wiki/Communicating_and_getting_help"
BUG_REPORT_URL="https://bugzilla.redhat.com/"
REDHAT_BUGZILLA_PRODUCT="Fedora"
REDHAT_BUGZILLA_PRODUCT_VERSION=32
REDHAT_SUPPORT_PRODUCT="Fedora"
REDHAT_SUPPORT_PRODUCT_VERSION=32
PRIVACY_POLICY_URL="https://fedoraproject.org/wiki/Legal:PrivacyPolicy"
VARIANT="Workstation Edition"
VARIANT_ID=workstation
Fedora release 32 (Thirty Two)
Fedora release 32 (Thirty Two)
```

**Output of `uname -a`:**

```
Linux fedora 5.6.6-300.fc32.x86_64 #1 SMP Tue Apr 21 13:44:19 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
```

**Output of `cat ~/.config/containers/storage.conf`:**
```
[storage]
driver = "vfs"
rootless_storage_path = "$HOME/.local/share/containers/storage"
[storage.options]
[storage.options.vfs]
ignore_chown_errors = "true"
```

**Output of `/etc/containers/storage.conf`:**
```
[storage]
driver = "vfs"
runroot = "/var/run/containers/storage"
graphroot = "/var/lib/containers/storage
rootless_storage_path = "$HOME/.local/share/containers/storage"
[storage.options]
[storage.options.vfs]
ignore_chown_errors = "true"
```

**Output of `BUILDAH_ISOLATION=crun buildah --log-level=debug info`:**
```
WARN error reading allowed ID mappings: error reading subuid mappings for user "mdhillon" and subgid mappings for group "mdhillon": No subuid ranges found for user "mdhillon" in /etc/subuid
WARN Found no UID ranges set aside for user "mdhillon" in /etc/subuid.
WARN Found no GID ranges set aside for user "mdhillon" in /etc/subgid.
DEBU running [buildah-in-a-user-namespace --log-level=debug info] with environment [BUILDAH_RUNTIME=crun SHELL=/bin/bash HISTCONTROL=ignoredups HISTSIZE=1000 HOSTNAME=fedora PWD=/home/mdhillon/docker LOGNAME=mdhillon XDG_SESSION_TYPE=tty HOME=/home/mdhillon LANG=en_US.UTF-8 LS_COLORS=rs=0:di=38;5;33:ln=38;5;51:mh=00:pi=40;38;5;11:so=38;5;13:do=38;5;5:bd=48;5;232;38;5;11:cd=48;5;232;38;5;3:or=48;5;232;38;5;9:mi=01;37;41:su=48;5;196;38;5;15:sg=48;5;11;38;5;16:ca=48;5;196;38;5;226:tw=48;5;10;38;5;16:ow=48;5;10;38;5;21:st=48;5;21;38;5;15:ex=38;5;40:*.tar=38;5;9:*.tgz=38;5;9:*.arc=38;5;9:*.arj=38;5;9:*.taz=38;5;9:*.lha=38;5;9:*.lz4=38;5;9:*.lzh=38;5;9:*.lzma=38;5;9:*.tlz=38;5;9:*.txz=38;5;9:*.tzo=38;5;9:*.t7z=38;5;9:*.zip=38;5;9:*.z=38;5;9:*.dz=38;5;9:*.gz=38;5;9:*.lrz=38;5;9:*.lz=38;5;9:*.lzo=38;5;9:*.xz=38;5;9:*.zst=38;5;9:*.tzst=38;5;9:*.bz2=38;5;9:*.bz=38;5;9:*.tbz=38;5;9:*.tbz2=38;5;9:*.tz=38;5;9:*.deb=38;5;9:*.rpm=38;5;9:*.jar=38;5;9:*.war=38;5;9:*.ear=38;5;9:*.sar=38;5;9:*.rar=38;5;9:*.alz=38;5;9:*.ace=38;5;9:*.zoo=38;5;9:*.cpio=38;5;9:*.7z=38;5;9:*.rz=38;5;9:*.cab=38;5;9:*.wim=38;5;9:*.swm=38;5;9:*.dwm=38;5;9:*.esd=38;5;9:*.jpg=38;5;13:*.jpeg=38;5;13:*.mjpg=38;5;13:*.mjpeg=38;5;13:*.gif=38;5;13:*.bmp=38;5;13:*.pbm=38;5;13:*.pgm=38;5;13:*.ppm=38;5;13:*.tga=38;5;13:*.xbm=38;5;13:*.xpm=38;5;13:*.tif=38;5;13:*.tiff=38;5;13:*.png=38;5;13:*.svg=38;5;13:*.svgz=38;5;13:*.mng=38;5;13:*.pcx=38;5;13:*.mov=38;5;13:*.mpg=38;5;13:*.mpeg=38;5;13:*.m2v=38;5;13:*.mkv=38;5;13:*.webm=38;5;13:*.webp=38;5;13:*.ogm=38;5;13:*.mp4=38;5;13:*.m4v=38;5;13:*.mp4v=38;5;13:*.vob=38;5;13:*.qt=38;5;13:*.nuv=38;5;13:*.wmv=38;5;13:*.asf=38;5;13:*.rm=38;5;13:*.rmvb=38;5;13:*.flc=38;5;13:*.avi=38;5;13:*.fli=38;5;13:*.flv=38;5;13:*.gl=38;5;13:*.dl=38;5;13:*.xcf=38;5;13:*.xwd=38;5;13:*.yuv=38;5;13:*.cgm=38;5;13:*.emf=38;5;13:*.ogv=38;5;13:*.ogx=38;5;13:*.aac=00;36:*.au=00;36:*.flac=00;36:*.m4a=00;36:*.mid=00;36:*.midi=00;36:*.mka=00;36:*.mp3=00;36:*.mpc=00;36:*.ogg=00;36:*.ra=00;36:*.wav=00;36:*.oga=00;36:*.opus=00;36:*.spx=00;36:*.xspf=00;36: SSH_CONNECTION=10.211.55.2 53634 10.211.55.12 22 XDG_SESSION_CLASS=user TERM=xterm-256color LESSOPEN=||/usr/bin/lesspipe.sh %s USER=mdhillon SHLVL=1 XDG_SESSION_ID=3 XDG_RUNTIME_DIR=/run/user/1000 SSH_CLIENT=10.211.55.2 53634 22 XDG_DATA_DIRS=/home/mdhillon/.local/share/flatpak/exports/share:/var/lib/flatpak/exports/share:/usr/local/share:/usr/share PATH=/home/mdhillon/.local/bin:/home/mdhillon/bin:/usr/local/bin:/usr/bin:/usr/local/sbin:/usr/sbin DBUS_SESSION_BUS_ADDRESS=unix:path=/run/user/1000/bus MAIL=/var/spool/mail/mdhillon SSH_TTY=/dev/pts/0 OLDPWD=/home/mdhillon _=/usr/bin/buildah TMPDIR=/var/tmp _CONTAINERS_USERNS_CONFIGURED=1 BUILDAH_ISOLATION=rootless], UID map [{ContainerID:0 HostID:1000 Size:1}], and GID map [{ContainerID:0 HostID:1000 Size:1}]
DEBU [graphdriver] trying provided driver "vfs"
DEBU vfs: ignore_chown_errors=true
{
    "host": {
        "CgroupVersion": "v1",
        "Distribution": {
            "distribution": "fedora",
            "version": "32"
        },
        "MemTotal": 11975348224,
        "MenFree": 10125873152,
        "OCIRuntime": "crun",
        "SwapFree": 5595197440,
        "SwapTotal": 5595197440,
        "arch": "amd64",
        "cpus": 4,
        "hostname": "fedora",
        "kernel": "5.6.6-300.fc32.x86_64",
        "os": "linux",
        "rootless": true,
        "uptime": "1h 29m 26.88s (Approximately 0.04 days)"
    },
    "store": {
        "ContainerStore": {
            "number": 1
        },
        "GraphDriverName": "vfs",
        "GraphOptions": [
            "vfs.ignore_chown_errors=true"
        ],
        "GraphRoot": "/home/mdhillon/.local/share/containers/storage",
        "GraphStatus": {},
        "ImageStore": {
            "number": 1
        },
        "RunRoot": "/run/user/1000/containers"
    }
}
```

**Output of `crun --version`:**
```
crun version 0.14.1
commit: 598ea5e192ca12d4f6378217d3ab1415efeddefa
spec: 1.0.0
+SYSTEMD +SELINUX +APPARMOR +CAP +SECCOMP +EBPF +YAJL
```

**Comment by mrinaldhillon** [comment]
@rhatdan could you please comment on this. Thanks

**Comment by rhatdan** [comment]
These are working for me with the overlay driver?  Why are you using the VFS Driver?

**Comment by rhatdan** [comment]
Very interesting.  I don't know why. @giuseppe @nalind Any ideas?

**Comment by rhatdan** [comment]
I got similar errors with Overlay.

**Comment by giuseppe** [comment]
the error from the OCI runtime:
```
2020-07-14T20:42:01.000174777Z: mount `devpts` to '/dev/pts': Invalid argument
```

happens because buildah specifies gid=5 in the mount options, but there is no gid=5 with a single mapping.  Podman handles this case and drops the gid=5 if there are not enough mappings.


The other issue with setgroups happens because the namespace was configured with a single mapping writing directly to `/proc/$PID/*id_map` and for doing that it is necessary to set `deny` into `/proc/$PID/setgroups`

**Comment by giuseppe** [comment]
PR here: https://github.com/containers/buildah/pull/2468

The issue with the runtime instead should be solved using `--isolation=rootless`

**Comment by mrinaldhillon** [comment]
@giuseppe @rhatdan 

Will this enable podman build to run with single uid mapping and chroot isolation without need of runc or crun?

**Comment by giuseppe** [comment]
> Will this enable podman build to run with single uid mapping and chroot isolation without need of runc or crun?

yes, you should be able to use podman build without an OCI runtime

**Comment by giuseppe** [comment]
I think we can close the issue since the PR was merged

**Comment by mrinaldhillon** [comment]
@rhatdan can we get a minor release with this fix?

**Comment by rhatdan** [comment]
@TomSweeneyRedHat Lets cut a new release.

**Comment by TomSweeneyRedHat** [comment]
@rhatdan 1.15.1 against master I presume?

**Comment by nalind** [comment]
I think making a 1.15.1 would involve cherry-picking the desired fix onto `release-1.15` and cutting the release from the branch.

**Comment by rhatdan** [comment]
If we have no new features, then we can do this. Also do we want these fixes in RHEL8.3?  
If there are new features we should cut 1.16.0 And RHEL8.3.1 will get the new version.
